src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,600 @@ ep_rt_mono_write_event_ee_startup_start (void)
+//
+// Outermost level of ETW-type-logging.  This method is used to log a unique type identifier
+// (in this case a MonoType) and (recursively) its type parameters when present.
+//
+// Arguments:
+//	* p_type_logger - BulkTypeEventLogger instance
+//      * mono_type - MonoType to be logged
+//      * type_id - Unique type identifier of mono_type
+//
+// This follows CoreCLR's BulkTypeEventLogger::LogTypeAndParameters
+
+void
+ep_rt_mono_log_type_and_parameters_if_necessary (
+	BulkTypeEventLogger *p_type_logger,
+	MonoType *mono_type,
+	intptr_t type_id)
+{
+	// TODO Log the type if necessary
+
+	ep_rt_mono_log_type_and_parameters (p_type_logger, mono_type, type_id);
+}
+
+// ETW has a limit for maximum event size. Do not log overly large method type argument sets
+static const uint32_t K_MAX_METHOD_TYPE_ARGUMENT_COUNT = 1024;
+
+//---------------------------------------------------------------------------------------
+//
+// ep_rt_mono_send_method_details_event is the method responsible for sending details of
+// methods involved in events such as JitStart, Load/Unload, Rundown, R2R, and other
+// eventpipe events. It calls ep_rt_mono_log_type_and_parameters_if_necessary to log
+// unique types from the method type and available method instantiation parameter types
+// that are ultimately emitted as a BulkType event in ep_rt_mono_fire_bulk_type_event.
+// After appropraitely logging type information, it sends method details outlined by
+// the generated dotnetruntime.c and ClrEtwAll manifest.
+//
+// Arguments:
+//      * method - a MonoMethod hit during an eventpipe event
+//
+// This follows CoreCLR's ETW::MethodLog::SendMethodDetailsEvent
+
+void
+ep_rt_mono_send_method_details_event (MonoMethod *method)
+{
+	if (method->wrapper_type != MONO_WRAPPER_NONE || method->dynamic)
+		return;
+
+	MonoGenericContext *method_ctx = mono_method_get_context (method);
+
+	MonoGenericInst *method_inst = NULL;
+	if (method_ctx)
+		method_inst = method_ctx->method_inst;
+
+	if (method_inst && method_inst->type_argc > K_MAX_METHOD_TYPE_ARGUMENT_COUNT)
+		return;
+
+	BulkTypeEventLogger *p_type_logger = ep_rt_bulk_type_event_logger_init ();
+
+	intptr_t method_type_id = 0;
+	g_assert (mono_metadata_token_index (method->token) != 0);
+	uint32_t method_token = mono_metadata_make_token (MONO_TABLE_METHOD, mono_metadata_token_index (method->token));
+	uint64_t loader_module_id = 0;
+	MonoClass *klass = method->klass;
+	if (klass) {
+		MonoType *method_mono_type = m_class_get_byval_arg (klass);
+		method_type_id = get_typeid_for_class (klass);
+
+		ep_rt_mono_log_type_and_parameters_if_necessary (p_type_logger, method_mono_type, method_type_id);
+
+		loader_module_id = (uint64_t)mono_class_get_image (klass);
+	}
+
+	uint32_t method_inst_parameter_types_count = 0;
+	if (method_inst)
+		method_inst_parameter_types_count = method_inst->type_argc;
+
+	intptr_t method_inst_parameters_type_ids [K_MAX_METHOD_TYPE_ARGUMENT_COUNT];
```

Comment:
This will bring in a rather large stack allocation (8K on 64-bit) in potential deep callstacks, not sure we should do anything about it now, just wanted to call it out.

Reply:
Right, there were complaints on CI about `method_inst_parameter_types_count` being variable at compile time. I just did this to make it a constant, and technically, we accept that many parameters in the method inst `if (method_inst && method_inst->type_argc > K_MAX_METHOD_TYPE_ARGUMENT_COUNT)`. Is there another way to keep it a constant at compile time? I'm now realizing that the type on CoreCLR is a `StackSArray`, which if its like the `StackSString` type, it will be a fixed size until it needs to allocate more memory.

Would we prefer to implement that sort of type?

Reply:
One option for this is to use g_alloca, dynamic stack allocation, that way you can alloc needed bytes on stack, just make sure you don't allocate to much (keep the limit of max method) and that you don't do it in any loops, alloca is not scoped on all platforms so doing it loops will just increase stack usage util function exists. We already use g_alloca in other parts of the runtime.

Reply:
As discussed separately, we are now using a dedicated mempool for the BulkTypeEventLogger to assist in dynamically allocating enough memory for the relevant type parameters.